pkg/ccl/changefeedccl/changefeed_processors.go

```diff
@@ -1175,56 +1198,85 @@ func (cf *changeFrontier) noteResolvedSpan(d rowenc.EncDatum) error {
 		return err
 	}
+
 	isBehind := cf.maybeLogBehindSpan(frontierChanged)
-	if frontierChanged {
-		if err := cf.handleFrontierChanged(isBehind); err != nil {
+
+	// Checkpoint job record progress if needed.
+	// NB: Sinkless changefeeds will not have a job state (js). In fact, they
+	// have no distributed state whatsoever. Because of this they also do not
+	// use protected timestamps.
+	if cf.js != nil {
+		if err := cf.maybeCheckpointJob(frontierChanged, isBehind); err != nil {
 			return err
 		}
 	}
+
+	if frontierChanged {
+		// Keeping this after the checkpointJobProgress call will avoid
+		// some duplicates if a restart happens.
+		newResolved := cf.frontier.Frontier()
+		cf.metrics.mu.Lock()
+		if cf.metricsID != -1 {
+			cf.metrics.mu.resolved[cf.metricsID] = newResolved
+		}
+		cf.metrics.mu.Unlock()
+
+		return cf.maybeEmitResolved(newResolved)
+	}
+
 	return nil
 }
 
-func (cf *changeFrontier) handleFrontierChanged(isBehind bool) error {
-	newResolved := cf.frontier.Frontier()
-	cf.metrics.mu.Lock()
-	if cf.metricsID != -1 {
-		cf.metrics.mu.resolved[cf.metricsID] = newResolved
-	}
-	cf.metrics.mu.Unlock()
+func (cf *changeFrontier) maybeCheckpointJob(frontierChanged, isBehind bool) error {
+	// Update checkpoint if the frontier has not changed, but it is time to checkpoint.
+	// If the frontier has changed, we want to write an empty checkpoint record indicating
+	// that all spans have reached the frontier.
+	updateCheckpoint := !frontierChanged && cf.js.canCheckpointFrontier()
 
-	checkpointStart := timeutil.Now()
-	if err := cf.checkpointResolvedTimestamp(newResolved, isBehind); err != nil {
-		return err
+	var checkpoint jobspb.ChangefeedProgress_Checkpoint
+	if updateCheckpoint {
+		cf.frontier.checkpoint(&checkpoint)
 	}
-	cf.metrics.CheckpointHistNanos.RecordValue(timeutil.Since(checkpointStart).Nanoseconds())
 
-	if err := cf.maybeEmitResolved(newResolved); err != nil {
-		return err
+	if frontierChanged || updateCheckpoint {
+		cf.js.lastFrontierCheckpoint = timeutil.Now()
+		checkpointStart := timeutil.Now()
+		if err := cf.checkpointJobProgress(
+			cf.frontier.Frontier(), frontierChanged, checkpoint, isBehind,
+		); err != nil {
+			return err
+		}
+		cf.metrics.CheckpointHistNanos.RecordValue(timeutil.Since(checkpointStart).Nanoseconds())
 	}
+
 	return nil
 }
 
-// checkpointResolvedTimestamp checkpoints a changefeed-level resolved timestamp
-// to the jobs record. It additionally manages the protected timestamp state
-// which is stored in the job progress details. It is only called if the new
-// resolved timestamp is later than the current one. The isBehind argument is
-// used to determine whether an existing protected timestamp should be released.
-func (cf *changeFrontier) checkpointResolvedTimestamp(
-	resolved hlc.Timestamp, isBehind bool,
+// checkpointJobProgress checkpoints a changefeed-level job information.
+// In addition, f 'manageProtected' is true, which only happens when frontier advanced,
```

Comment:
```suggestion
// In addition, if 'manageProtected' is true, which only happens when frontier advanced,
```

Reply:
Yes; 10MB is scary.  Question is: another PR to specifically target this deficiency vs this pr...